### ISSUE TITLE
CiviCRM 4.7 not working under BOA

### ISSUE THREAD

**Issue by sluc23** [omega8cc/boa#1223]
We are trying to use CiviCRM 4.7 under BOA and there is a very specific bug that is blocking us.
It is a BOA related issue, because CiviCRM 4.7 works fine under other environments, including AEgir standalone context.
Performing some tasks from the CiviCRM Web UI (like Refreshing/Installing extensions), leads to a php error:

`
[25-Oct-2017 15:42:09 UTC] PHP Fatal error:  require(): Failed opening required '<platform_path>/sites/civicrm47.mysite.org/files/civicrm/templates_c/CachedExtLoader.a01984b452befc79625bfc5ff3215d7a.php' (include_path='.:<platform_path>/profiles/my_profile/modules/contrib/civicrm/..//civicrm:<platform_path>/profiles/my_profile/modules/contrib/civicrm/..//civicrm/packages:.:<platform_path>/profiles/my_profile/modules/contrib/civicrm:<platform_path>/profiles/my_profile/modules/contrib/civicrm/packages:<platform_path>/profiles/my_profile/modules/contrib/civicrm/vendor/tecnickcom:<platform_path>/profiles/my_profile/modules/contrib/civicrm/vendor/php in <platform_path>/profiles/my_profile/modules/contrib/civicrm/CRM/Extension/ClassLoader.php on line 86
`

There is something weird with that file, which in most of the cases exists, but cannot be opened.. my guess is that must be a config on templates_c dir that avoids accesing from php via web (same php scripts executed from CLI works fine)

This issue was reported in CiviCRM issue tracker too -> [CRM-19781](https://issues.civicrm.org/jira/browse/CRM-19781?focusedCommentId=109301&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-109301)

We are researching the cause of it, just reporting it here in case someone else has faced it too

[boa info](https://gist.github.com/sluc23/45cc7aea817c046925ac7146d8fd3d77)

**Comment by omega8cc** [comment]
Can you try to switch from `_PERMISSIONS_FIX=NO` to `_PERMISSIONS_FIX=YES` and run `bash /var/xdrago/daily.sh` script?

It should fix permissions on existing files, but it may still not help with files you will re-create again via Web UI, because these files will be owned by very restricted web server system user, and the it may cause problems if they are created with incorrect permissions by CiviCRM itself.

**Note: You can also fix permissions just by running Verify task on the affected site.**

It's also not clear if this happens only with PHP 7.0, but works with PHP 5.6 ?

**Comment by omega8cc** [comment]
Also, it should help if you could let us know the result of this command directly after running the task via CiviCRM Web UI:

`ls -la <platform_path>/sites/example.com/files/civicrm/templates_c/CachedExtLoader.xyz.php`
`ls -la <platform_path>/sites/example.com/files/civicrm/templates_c`

**Comment by sluc23** [comment]
switch to` _PERMISSIONS_FIX=YES` and ran bash `/var/xdrago/daily.sh`
Same error :(

`ls -la <platform_path>/sites/example.com/files/civicrm/templates_c`

```
o2.ftp:[civicrm47.mysite.org]$ ls -la files/civicrm/templates_c/
total 36
drwxrwsr-x  3 o2     www-data  4096 Oct 30 15:34 .
drwxrwsr-x  9 o2     www-data  4096 Oct 25 11:13 ..
-rw-rw-r--  1 o2.web www-data    56 Oct 30 15:33 .htaccess
-rw-rw-r--  1 o2.web www-data 20444 Oct 30 15:33 CachedCiviContainer.a01984b452befc79625bfc5ff3215d7a.php
drwxrwsr-x 82 o2     www-data  4096 Oct 30 15:33 en_US
```

For some reason this [php script](https://github.com/civicrm/civicrm-core/blob/master/CRM/Extension/ClassLoader.php#L91) cannot create the file.. or if it's created, then somehow it's erased..

I'm going to debug step-by-step this code to check if I find any issue

**Comment by omega8cc** [comment]
Hmm.. the permissions look correct, and web server should be able to easily read it. Have you tried to reproduce this with PHP 5.6? Maybe something is broken in our PHP7 config?

**Comment by sluc23** [comment]
Yes I have same behavior on PHP 5.6 or PHP 7.0.
When using multi-fpm.info to set PHP .70 to a site, is it normal to have this user as the web one? 

**o2.70.web**

**Comment by sluc23** [comment]
Well after long time debugging I couldn't find the cause why this issue is happening.
Anyone has managed to have CiviCRM 4.7 working in BOA?

specially when trying to install and/or work with civicrm's extensions

**Comment by sluc23** [comment]
This CiviCRM's patch seems to work, when the file exists but the user has no perms to open it (in some very specific conditions)
Not sure the implications of recreating caches files when they already exist, but at least is the first step I've made to have CiviCRM 4.7 working on BOA so far.
I'll keep on debugging anyway

https://gist.github.com/sluc23/cb491aff6fdb3eee224045ead30e40f9

**Comment by omega8cc** [comment]
Hmm.. thanks for the feedback, but obviously we need to fix any related problems in Aegir and/or BOA, without requiring patches for CiviCRM.

Can you confirm that you didn't try to install/add CiviCRM on an already installed Drupal site?

Also, can you provide more details on:

> specially when trying to install and/or work with civicrm's extensions

Thanks!

**Comment by sluc23** [comment]
I've always tested in a site where CiviCRM is installed when the site is created, never installed CiviCRM manually.
CiviCRM 4.7 installation finishes without an issue, most of the actions/tasks work fine as far as I've tested them, but I've got the error in these scenarios:
1. cleaning CiviCRM cache's repeatedly, excuting this url at least 2/3 times in a row (/civicrm/clearcache) 
2. refreshing CiviCRM Extension's list (/civicrm/admin/extensions?reset=1 - Refresh button) 
3. time to time got some random WSOD performing some tasks, but couldn't figure it out steps to reproduce this, I guess it is related to clear cache too

First using the default system's php (5.6) got WSOD without any specific error, then switched to multi-fpm and with both 5.6 and 7.0 got same error:

```
[22-Feb-2018 13:31:54 UTC] PHP Fatal error:  require(): Failed opening required '<drupal_root>/files/civicrm/templates_c/CachedExtLoader.f5813a76d27f11433fd09e1ebc2ec5f8.php' (include_path='.:/ .. .. ../ in /data/disk/o1/static/civicrm/civicrm-7.56-4.7.30/profiles/civicrm_profile/modules/contrib/civicrm/CRM/Extension/ClassLoader.php on line 92
```
or
```
[22-Feb-2018 13:33:17 UTC] PHP Fatal error:  require_once(): Failed opening required '<drupal_root>/files/civicrm/templates_c/CachedCiviContainer.f5813a76d27f11433fd09e1ebc2ec5f8.php' (include_path='.:/ .. .. ../ in /data/disk/o1/static/civicrm/civicrm-7.56-4.7.30/profiles/civicrm_profile/modules/contrib/civicrm/Civi/Core/Container.php on line 90
```

These 2 "cache" files are used by new "Autoloaders" classes added in CiviCRM 4.7, and the weird thing I've found is that php function **file_exists($file)** returns true but then the file cannot be opened, so I've added **fileperms($file)** and in some cases it returns **0**
That's why added this **fileperms($file)** in my patch in if condition it would avoid to try to read a file when php has no permissions on it, and enters into else to recreate it.

**Comment by omega8cc** [comment]
We have tried to debug this and here what we have found so far:

1. There is no configuration in place which would prevent including files from `templates_c`
2. There are no permissions issues, because the file and it's parent dir are fully writable/readable
3. We do block web access to `templates_c` but it's expected by CiviCRM (see .htaccess file)
4. We can't disable `open_basedir` restrictions, but they are not relevant here anyway
5. Installing extensions is a different topic we can discuss

Our conclusion at the moment is that some random errors happening when CiviCRM complains that it can't include a file which actually exists and is readable by the PHP engine, are CiviCRM problems with managing cache stampede syndrome.

We appreciate that you don't experience these symptoms on non-BOA servers with wide open configuration and no protection in place, however we still can't find what exactly makes recent CIviCRM version think that the environment is "broken", and solution certainly can't rely on disabling all kinds of system protection. 

We should also perhaps make a list of actual issues which can be considered as deal breakers and others which are just some annoying but actually harmless symptoms.

Clearing the CiviCRM cache is a good example: CiviCRM deletes the files from `templates_c` and immediately checks if they exist (and complains) instead of recreate them first and *then* check if they exist. This doesn't make any sense. You can see this absurd behaviour if you click the "Cleanup Caches" button, the page reloads and you get WSOD or error inline if you enabled debugging mode, because CiviCRM tried to include file it deleted but didn't recreate yet. A classic chicken/egg problem on the app level. That WSOD/first request did recreate the file then, so when you refresh the page in your browser, the error is gone. It's not server's fault, thought, but application logic fault.

Next, managing codebase (extensions) via web UI is convenient but dangerous, if allowed. CiviCRM is using Drupal files space as a workaround, so it may work, but downloading will not work most of the time, for reasons explained in this thread: https://forum.civicrm.org/index.php%3Ftopic=35066.0.html

At this moment we suspect that these problems are a result of CiviCRM bugs which are exposed when used with BOA stack, and not a result of too restrictive configuration in BOA, because there is literally nothing which could be identified as a culprit on the BOA side -- we have spent literally days on careful tests and attempts to replicate the problems to make sure what actually happens.

Obviously, we are very interested to make it work, but pointing out that it works just fine on a generic hosting doesn't offer any meaningful arguments, especially if we have thoroughly checked and confirmed that there are no restrictions or limitations which could cause such weird problems.

It's very possible that the end result will be a patch for CiviCRM to improve its internal logic, but we need to find which of its pseudo-tests (like complaining for no good reason about `open_basedir`) trigger the actual errors, like by including a file too early after it was deleted, and before it was recreated, without any lock and double check in place, etc.

Next steps in debugging will include carefully disabling all standard environment protection measures in PHP configuration, until we will determine where exactly and how CiviCRM fails.

In the meantime can we focus here on creating a list of things which affect website stability or its management via Clone and Migrate tasks etc, and those which are just random annoyances etc. so we could figure out the actual scale of the problem to solve?

Thank you for your feedback!

**Comment by sluc23** [comment]
thanks @omega8cc for your analysis.

I tried your recommendation, so I apply ```is_readable()``` and I've got the same php error as using ```file_exists()``` using this [patch](https://gist.github.com/sluc23/4188ae5c1432872b34813376ce727f49) in php 7.0, and **Cleanup Caches** button

``` PHP Fatal error:  require(): Failed opening required '/...../<site>/files/civicrm/templates_c/CachedExtLoader.4e552eb05b8e410117e795112a7de946.php' . . . . . ```

```is_readable()``` returns **TRUE**, but then ```$file``` cannot be included. 
*BTW: Cleaning caches through CiviCRM's api works fine ```drush cvapi system.flush ``` without any patch*

I agree that CiviCRM core might have some bugs/bad practices in code that it could lead to issues, but my concern is if we cannot rely on functions like ```file_exists()```, ```is_file()```, ```is_readable()```, etc to get values and fix the incorrect logic, it will be hard to make a patch to fix these issues and call CiviCRM's community attention to get it included in core.

At this point the issues that I've found are in CiviCRM 5.0:
1. Cleanup Caches
2. Refresh Extension's list

I woudn't label both issues as **critical**, but as **important** in a typical CiviCRM's usage

And as I mentioned these 2 issues seem to be fixed in PHP 7.0 with these [patches](https://gist.github.com/sluc23/cb491aff6fdb3eee224045ead30e40f9) but still, I'm not confident enough to roll this up in prod until we are more certain about the real cause of it

**Comment by omega8cc** [comment]
Yeah, that is expected. While `is_readable()` should be used there, it doesn't fix the visible cache stampede syndrome, which needs to be fixed in CiviCRM with proper lock logic.

It's also possible that it doesn't take into account opcache, but the lack of proper locking during cache clear is the actual problem here. You don't experience the problem on command line because there is no PHP caching of any kind on CLI, perhaps.

Again, it's not that we can't count on `file_exists()`, because we can. The problem is that CiviCRM is not using proper lock on cache clear.

I don't think it will be hard to fix, it's actually trivial and obvious thing to handle in any PHP app. I have seen Grav CMS doing the same bad thing, BTW, which causes problems on upgrades and any cache clears for exactly the same reason.

The patches you have linked are only temporary workarounds hiding the culprit, while it needs to be fixed properly.

While we are at it, we have extensively tested 4.6, 4.7 and 5.x CiviCRM with both Drupal 7 and Drupal 6, and found some regressions we didn't see before, which need to be fixed on the Aegir side too.

A bit of rabbit hole, I must admit, but we are determined to find and fix all issues reported and discovered.

**Comment by sluc23** [comment]
hi @omega8cc, were you able to make any progress on this issue? tx

**Comment by omega8cc** [comment]
Yes, we have run many tests, with many CiviCRM and Drupal core versions, and we believe that we have found the culprit and reliable workaround is now tested, without patching CiviCRM, which is important. Once we complete all tests and fix unavoidable regressions this may cause for non-CiviCRM sites and apps, we will share and link it here, hopefully in the next few days. Stay tuned!

**Comment by omega8cc** [comment]
Problem solved and tested with CiviCRM up to 5.3.0 -- we have dropped attempts to introduce complex workarounds we have tested before after discovering that we have had duplicate Zend OPcache config directives on php.ini and nginx level, which introduced confusion during testing and debugging. This cleanup improved also the dev/debug mode: https://github.com/omega8cc/boa/commit/8f7b4948ed17159b2bfa29fd6b9706638296715c

**Comment by sluc23** [comment]
@omega8cc great news!! 
This will be part of BOA 3.3.0 right?, any ETA for releasing this?

thanks for your hard work :)

**Comment by omega8cc** [comment]
Yes, it will be included in the upcoming 3.3.0 stable, which is planned for release in about two weeks. We will move as many tasks still in progress to 3.3.1 and then 3.4.0 as soon as possible, so we could speed up things considerably this month and not delay 3.3.0 any longer.
